pkg/webhook/mutation.go

```diff
@@ -0,0 +1,158 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/open-policy-agent/cert-controller/pkg/rotator"
+	opa "github.com/open-policy-agent/frameworks/constraint/pkg/client"
+	"github.com/open-policy-agent/gatekeeper/apis"
+	"github.com/open-policy-agent/gatekeeper/pkg/controller/config/process"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation"
+	"github.com/open-policy-agent/gatekeeper/pkg/util"
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	clientcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+type mutationResponse string
+
+const (
+	mutationSkipResponse    mutationResponse = "skip"
+	mutationUnknownResponse mutationResponse = "unknown"
+)
+
+func init() {
+	AddToManagerFuncs = append(AddToManagerFuncs, AddMutatingWebhook)
+
+	if err := apis.AddToScheme(runtimeScheme); err != nil {
+		log.Error(err, "unable to add to scheme")
+		panic(err)
+	}
+}
+
+// +kubebuilder:webhook:verbs=create,path=/v1/mutate,mutating=true,failurePolicy=ignore,groups=*,resources=*,versions=*,name=mutation.gatekeeper.sh
+// +kubebuilder:rbac:groups=*,resources=*,verbs=get;list;watch;update
+
+// AddMutatingWebhook registers the mutating webhook server with the manager
+func AddMutatingWebhook(mgr manager.Manager, client *opa.Client, processExcluder *process.Excluder) error {
```

Comment:
Should we add pkg/webhook/mutation_test.go now or later?

Reply:
I added some tests to https://github.com/open-policy-agent/gatekeeper/pull/978 , specifically for the validation part.
I'd say it makes sense to add the rest when everything is wired up and the webhook is able to mutate.

Reply:
Can we add an issue to the mutation milestone for adding webhook tests?

Reply:
Added #995
Please add a mutation label

Reply:
Thanks! added